### ISSUE TITLE
[UI components] Documentation pages have duplicate ids

### ISSUE THREAD

**Issue by shawnbot** [18F/web-design-standards#1225]
## Description

The accordion documentation is a prime example because its demo HTML presents an accordion with the same ids as the "Code" and "Documentation" collapsibles below (they both use the `collapsible-n` naming convention), but all of the component pages exhibit the same issue. 
- The [styleguide.js](https://github.com/18F/web-design-standards/blob/staging/docs/doc_assets/js/styleguide.js#L34) script inserts a `collapsible-0` id for the "Code" area on all of the component pages, which is the same as the "Documentation" area's id.
- The "Code" and "Documentation" areas should be given more meaningful ids (e.g. `code` and `documentation`) so that you can also link directly to them on the page.
- There should be tests in place to help us avoid this problem. [html-proofer](https://github.com/gjtorikian/html-proofer) works great, but maybe there's a similar tool built in Node? [pa11y](http://pa11y.org/) would surface these issues, too.
## Steps to reproduce the issue
1. Visit: https://standards.usa.gov/accordions/
2. Run [HTML_CodeSniffer](http://squizlabs.github.io/HTML_CodeSniffer/).
3. Observe that duplicate ids are listed as an error.
4. Alternately, call `document.querySelectorAll('[id=collapsible-0]')` in the JS console to see the number of elements with the same id.
5. Install [ack](https://github.com/petdance/ack2) and run `ack -c collapsible-0 _site/**/*.html` on the generated site directory to see which files have the most duplicate ids of the first collapsible.

**Comment by shawnbot** [comment]
I've addressed the duplicate ids in the "Code" and "Documentation" collapsibles in #1224 by setting the former's id to `code` (instead of `collapsible-0`).

**Comment by juliaelman** [comment]
Fixed in #1224

**Comment by shawnbot** [comment]
@juliaelman I'm sorry for not being 100% clear about this. There are still lots of duplicate ids on some pages. I've only fixed the issue on the accordion docs page.

**Comment by joshbruce** [comment]
**Analysis results:** Pages using the `collapsible-0` attribute were the only pages with duplicate `id` attributes found. 

> What does `id="collapsible-0"` actually do? Is it deprecated and can it just be deleted?

A search for `collapsible-0` did not appear anywhere one might expect (as a selector for CSS and JS) - only the attribute definition within the element. This search also resulted in finding `collapsible-0` being used as an `aria-controls` value.

The analysis also turned up usage in combination with `aria-controls` attribute - but inconsistently applied. If the existence of `id="collapsible-0"` is an accessibility consideration (or requirement) - then `collapsible-0` could not be removed.

> Is the use of `id="collapsible-0"` meant to be an accessibility consideration? If so, documentation accordions, when there is more than one, would need unique identifiers and `aria-controls` attributes to correct the current inconsistency. Further, we'd probably want the other accordions (code mainly) to follow suit, yes?

--------------------
Checked denotes no duplicate `id` found.
Checked with notes indicates curiosity of whether the element actually `requires` an `id` for functional purposes.

- [x] Home
- [x] Getting started
- [ ] Visual style: duplicate `collapsible-0`.
- [x] Grids: `collapsible-0`
- [x] Buttons: `collapsible-0`
- [x] Labels: `collapsible-0`
- [x] Tables: `collapsible-0`
- [x] Alerts: `collapsible-0`
- [x] Accordions (thanks @shawnbot): `documentation` - not sure what this `id` does, if it became a norm, it could end up being in the same position as `collapsible-0`.
- [ ] Form controls: `collapsible-0`
- [ ] Form templates: `collapsible-0` - curious whether some of the `id` are actually being used for something. For our Standards we find some contributors adding attributes to elements because they believe *all* elements need to mimic something - not because the attributes are actually necessary - wonder if this is what happened with some of this `id` question.
- [x] Search bar: `collapsible-0`
- [x] Side navigation: `collapsible-0`
- [x] Footers: `collapsible-0`

Maybe we could create another issue regarding removing unnecessary `id` and `class` attributes. Add it to the "clean, semantic code` milestone as well...??

--------------------------
Since we're talking "clean, semantic code" - a recommendation about coding style - not sure how prescriptive we would like to be here.

I find HTML readability improves a great deal if `id` is first and precedes `class`, which precedes all other attributes. 

Because `id` must be unique to the page, it's easier to skim when `id` is in the first position because you don't have to sift through the other attributes applied to the element. See https://github.com/8fold/html5-generator-php - for the coding style recommendation I'm using right now for my more personal work on the score of attribute ordering.

**Comment by rogeruiz** [comment]
@juliaelman I've confirmed that @joshbruce's assessment above is correct. The only page with duplicate `id="collapsible-0"` are the form control and form template pages.

**Comment by shawnbot** [comment]
Thanks for the detailed assessment, @joshbruce! Glad to hear that there aren't that many left to fix.

**Comment by joshbruce** [comment]
@shawnbot - We need a h/t emoji, I think.

**Comment by shawnbot** [comment]
This is the best I can do: 👉 🎩

**Comment by maya** [comment]
>I find HTML readability improves a great deal if id is first and precedes class, which precedes all other attributes.

I follow the following attribute order in which class precedes id: 
http://codeguide.co/#html-attribute-order

>Classes make for great reusable components, so they come first. Ids are more specific and should be used sparingly (e.g., for in-page bookmarks), so they come second.

Although I don't think this guidance is in our coding styleguide yet.

**Comment by shawnbot** [comment]
Personally, I prefer id first because it makes it easier (for me, anyway) to scan the HTML and understand at a glance which elements are "important"—assuming that elements with an id are more "important" than those without. I'm happy to go with whatever convention y'all decide on, though!

**Comment by joshbruce** [comment]
@maya - Interesting and I do appreciate a lot of the style recommendations there. I can see the perspective. I think it is also a reason I prefer the `id` first as well. It hints to me, as a front-end developer, that chances are the `id` is not being used for styling for styling; instead, reference for some other purpose - aria labels, internal links, label `for` declarations, and so on. Classes on the other hand...??

Having said that, I'm with @shawnbot. 😉

**Comment by maya** [comment]
Here's more detail on what needs to be fixed:

**Typography** (9) 
- [ ] Duplicate id attribute value "collapsible-0" found on the web page. (7)
- [ ] Duplicate id attribute value "code" found on the web page. (2)

**Grids** (1)
- [ ] Duplicate id attribute value "code" found on the web page.

**Typography** (9) 
- [ ] Duplicate id attribute value "collapsible-0" found on the web page. (7)
- [ ] Duplicate id attribute value "code" found on the web page. (2)

**Form controls** (9)
- [ ] Duplicate id attribute value "collapsible-0" found on the web page. (4)
- [ ] Duplicate id attribute value "code" found on the web page. (4)

**Form templates** (7)
- [ ] Duplicate id attribute value "collapsible-0" found on the web page. (3)
- [ ] Duplicate id attribute value "code" found on the web page. (3)
- [ ] Duplicate id attribute value "password" found on the web page. (1)

**Comment by shawnbot** [comment]
There are no longer any duplicate ids on [accordions](https://standards.usa.gov/accordions/), but there are on [form controls](https://standards.usa.gov/form-controls/).

**Comment by shawnbot** [comment]
This was fixed in 18F/web-design-standards-docs#279.
